src/models/Order.js

```diff
@@ -0,0 +1,110 @@
+      ref: 'User', // Referencing the User model
+      required: [true, 'Please provide user'],
+    },
+    // The orderStatus field indicates the current status of the order
+    orderStatus: {
+      type: String,
+      enum: [
+        'pending',
+        'payment_successful',
+        'payment_failed',
+        'cancelled',
+        'complete',
+      ],
+      default: 'pending',
+      required: [true, 'Please provide an order status'], // Order status is required
+    },
+    // The tax field represents the tax percentage applied to the order
+    tax: {
+      type: Number,
+      required: [true, 'Please provide a tax percentage'], // Tax percentage is required
+      min: [0, 'Tax percentage must be at least 0'], // Tax percentage must be at least 0
+      max: [1, 'Tax percentage must be at most 1'], // Tax percentage must be at most 1
+    },
+    // The subtotal field represents the total price before taxes
+    subtotal: {
+      type: Number,
+      required: [true, 'Please provide a subtotal'],
+      min: [0, 'Subtotal must be at least 0'],
+    },
+    // The total field represents the total price after taxes
+    total: {
+      type: Number,
+      required: [true, 'Please provide a total'],
+      min: [0, 'Total must be at least 0'],
+    },
+    // The paymentIntentId field stores the Stripe state of the purchase
+    paymentIntentId: String, // Payment intent ID is a string
+    orderItems: [orderItemSchema],
+  },
+  {
+    timestamps: true, // Automatically add createdAt and updatedAt timestamps
+  }
+);
+
+// Create a function to update order statuses
+const updateOrderStatus = async () => {
+  try {
+    // Get the current date and subtract 1 minute
+    const oneMinutesAgo = new Date(Date.now() - 1 * 60 * 1000);
+
+    // Update orders with a status of "pending" that were created 1 minute ago or earlier,
+    // and set the order status to "cancelled"
+    const result = await Order.updateMany(
+      { orderStatus: 'pending', createdAt: { $lte: oneMinutesAgo } },
+      { $set: { orderStatus: 'cancelled' } }
+    );
+    console.log('Update result:', result);
+  } catch (error) {
+    console.error('Error updating orders:', error);
+  }
+};
+
+// Call the updateOrderStatus function every 1 minute (60 * 1000 milliseconds)
+setInterval(updateOrderStatus, 1 * 60 * 1000);
+
+// Middleware: Update order statuses before executing a find operation
+orderSchema.pre('find', async function (next) {
+  console.log('Pre-find started');
+  // Call the updateOrderStatus function to update order statuses
+  await updateOrderStatus();
+  console.log('Pre-find finished');
+  next();
+});
+
+// Middleware: Update order statuses before executing a findOne operation
+orderSchema.pre('findOne', async function (next) {
+  console.log('Pre-findOne started');
+  // Call the updateOrderStatus function to update order statuses
+  await updateOrderStatus();
+  console.log('Pre-findOne finished');
+  next();
+});
```

Comment:
Both of these worked for me as well! 

I created an order through the `POST /api/v1/orders` endpoint. Then, since there's not yet an endpoint for fetching orders, I instead opened a terminal window running node with all the `.env` variables: `node -r dotenv/config`

Then I ran the following in the node REPL to test that the above was working (I first commented out the `setInterval` just to make sure I was only testing the middlewares):  
 

```
> const connectDb = require('./src/db/connect')
undefined
> const db = await connectDb(process.env.MONGO_URL)
undefined

# Do a findById fetch (this should call the `pre-findone` middleware, but since 1 minute hasn't passed yet, nothing will be updated)
> await Order.Order.findById("64e1218480db8b4c9f15ff4b")
Pre-findOne started
Update result: {
  acknowledged: true,
  modifiedCount: 0,
  upsertedId: null,
  upsertedCount: 0,
  matchedCount: 0
}
Pre-findOne finished
{
  _id: new ObjectId("64e1218480db8b4c9f15ff4b"),
  user: new ObjectId("64d44bc8337399ccf9ad7e4d"),
  orderStatus: 'pending',
  tax: 0.14,
  subtotal: 50,
  total: 57,
  orderItems: [
    {
      album: new ObjectId("64d2a94c793389a43fc5a8ec"),
      quantity: 2,
      _id: new ObjectId("64e1218480db8b4c9f15ff4c")
    },
    {
      album: new ObjectId("64d2a94c793389a43fc5a8e7"),
      quantity: 5,
      _id: new ObjectId("64e1218480db8b4c9f15ff4d")
    }
  ],
  createdAt: 2023-08-19T20:09:40.331Z,
  updatedAt: 2023-08-19T20:09:40.735Z,
  __v: 0,
  paymentIntentId: 'pi_3NgvSqKnb0YvaiB30BpqeFkt'
}


# Now wait for 1 minute, and do the same thing again but this time the order will have been updated to 'cancelled':
> await Order.Order.findById("64e1218480db8b4c9f15ff4b")
Pre-findOne started
Update result: {
  acknowledged: true,
  modifiedCount: 1,
  upsertedId: null,
  upsertedCount: 0,
  matchedCount: 1
}
Pre-findOne finished
{
  _id: new ObjectId("64e1218480db8b4c9f15ff4b"),
  user: new ObjectId("64d44bc8337399ccf9ad7e4d"),
  orderStatus: 'cancelled',
  tax: 0.14,
  subtotal: 50,
  total: 57,
  orderItems: [
    {
      album: new ObjectId("64d2a94c793389a43fc5a8ec"),
      quantity: 2,
      _id: new ObjectId("64e1218480db8b4c9f15ff4c")
    },
    {
      album: new ObjectId("64d2a94c793389a43fc5a8e7"),
      quantity: 5,
      _id: new ObjectId("64e1218480db8b4c9f15ff4d")
    }
  ],
  createdAt: 2023-08-19T20:09:40.331Z,
  updatedAt: 2023-08-19T20:11:32.542Z,
  __v: 0,
  paymentIntentId: 'pi_3NgvSqKnb0YvaiB30BpqeFkt'
}
```

Reply:
setInterval is working for me, but code doesn't work without it, and even console.log doesn't work. I`m wondering where the mistake is.

Reply:
Hmm, what do you mean the code doesn't work without it? 

Without `setInterval` then the order status won't be automatically updated just based on time. However, the middlewares should still work; whenever a `Order.findById`, `Order.findOne`, or `Order.find` function is called. Are you observering a different outcome? Could you outline what steps you took to trigger the middleware (in the case without setInterval) to help me understand what you mean?

Reply:
I think I've found another solution. I'll make a commit so you can take a closer look. Using expiresAt in MongoDB instead of setInterval would automatically delete data when it becomes obsolete. This is more reliable and resource-efficient, especially for unpaid orders. Do you think this could be a better option? I can revert to the code with setInterval if this isn't a good idea in this case?